src/cmap/conn/stream.rs

```diff
@@ -38,27 +38,55 @@ pub(super) enum Stream {
     ),
 }
 
+fn try_connect(address: &SocketAddr, timeout: Duration) -> Result<TcpStream> {
+    // The URI options spec requires that the default connect timeout is 10 seconds, but that 0
+    // should indicate no timeout.
+    let stream = if timeout == Duration::from_secs(0) {
+        TcpStream::connect(address)?
+    } else {
+        TcpStream::connect_timeout(address, timeout)?
+    };
+
+    Ok(stream)
+}
+
+fn connect_stream(address: &StreamAddress, connect_timeout: Option<Duration>) -> Result<TcpStream> {
+    let timeout = connect_timeout.unwrap_or(DEFAULT_CONNECT_TIMEOUT);
+
+    let mut socket_addrs: Vec<_> = address.to_socket_addrs()?.collect();
+
+    if socket_addrs.is_empty() {
+        return Err(ErrorKind::NoDnsResults(address.clone()).into());
+    }
+
+    // After considering various approaches, we decided to do what other drivers do, namely try each
+    // of the addresses in sequence with a preference for IPv4.
+    socket_addrs.sort_by_key(|addr| if addr.is_ipv4() { 0 } else { 1 });
+
+    let mut connect_error = match try_connect(&socket_addrs[0], timeout) {
+        Ok(stream) => return Ok(stream),
+        Err(err) => err,
+    };
+
+    for address in &socket_addrs[1..] {
```

Comment:
could we combine the above connect into this loop? We could have `connect_error` start as `None` or something, then do an `unwrap_or_else` with `NoDnsResults`.